slides/week2/week2.org

```diff
@@ -0,0 +1,82 @@
+#+TITLE: Week 2
+#+AUTHOR: Kristaps Berzinch, Raymond Ortiz
+#+EMAIL: kristaps@robojackets.org, rortiz9@gatech.edu
+
+* What are we doing today?
+- Introduction to C++
+ - What it is
+ - Functions
+ - Includes and headers
+ - Working with strings
+- How to find and read documentation
+- Sending and recieving changes from remote Git repositories
+
+* What is C++?
+- A general-purpose programming language
+- Designed with performance, efficiency, and flexibility in mind
+- Great for robots!
+
+* Functions
+#+BEGIN_NOTES
+After introducing the concept, write a simple example. Don't go into prototypes yet, we'll cover them in a bit.
+
+TODO I think we should design a specific example that is easy to code on the fly but isn't completely trivial (e.g. writing an adder function seems dumb)
+#+END_NOTES
+- A function is a chunk of code with defined inputs and outputs
+ - Minimizes copy-and-paste
+ - Less buggy
+- These can be chained together for more complex functionality
+- You want each function to have a single high-level purpose
+
+* Includes
+#+BEGIN_NOTES
+After introducing the concept, demo some basic examples of including a file within another file. Still no prototypes.
+Include a function declaration in more than one file to cause a compilation error to lead into header files
+#+END_NOTES
+- Allows you to use code from other files
+- C++ has a lot of flexibility; this is a blessing and a curse
+- Your code will have little structure unless you organize it well
+- Break up your code logically and purposefully
+ - Readability above all
+
+* Header files
+#+BEGIN_NOTES
+Now you can talk about prototypes!
+#+END_NOTES
+- Only contain /declarations/, also known as prototypes, not definitions
+ - Function name
+ - Arguments
+ - Return type
+ - No functionality
+- This enables a lot of different files to "know" about a function but only have one definition for it
+
+* Working with strings
+#+BEGIN_NOTES
+Write and walk through some example code.
+#+END_NOTES
+- A C-style string is stored as an array of characters ending with a NULL (0x0)
+- The C++ standard library contains a string class with several useful functions, in =<string>=
+ - Use this instead of C-style unless you have a specific reason not to
+- Use the addition oprator (+) to concatenate strings
+- Use =to_string= to convert other types to strings
+- Use =sto___= functions to parse strings to other types
+| Name | Description |
+|-------+------|
+| =length= | Returns the length of a string |
+| =substr= | Returns a portion (substring) of the string |
+| =find= | Returns the position of a substring, if present |
+| =empty= | Returns true if length is zero, otherwise false |
+
+* How to find and read documentation
+#+BEGIN_NOTES
+Show how to get to the string documentation
+#+END_NOTES
+- Most of the last slide was sourced from cppreference.com
+- Documents standard library functionality
+- Great reference for all things C++

```

Comment:
Probably should have a link here to the reference homepage in the slides itself, so people can click it later.

http://en.cppreference.com/w/cpp/string/basic_string

Maybe even a link to the ddg bang: https://is.gd/sn8IQG

Also I tend to like cplusplus.com as well if you want something more consise: https://is.gd/nn5n43